### PR TITLE
SelectVariants works with non-diploids

### PR DESCRIPTION
**Pull request by ronlevine**
Implements https://github.com/broadinstitute/gatk/issues/1382.
Remove diploid assumptions during allele subsetting.

**Summary**
* Changed `GenotypeLikelihoods.GenotypeLikelihoodsAllelePair GenotypeLikelihoods.getAllelePair(int PLindex)` to `ArrayList<Integer> GenotypeLikelihoods.getAlleles(int PLindex, int ploidy)`  in GATKVariantContextUtils. 
* Added integration and unit tests.tests

### CODE REVIEW

**Comment by ronlevine** [comment]
@droazen Please review.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/SelectVariants.java [hunk 1]
```diff
@@ -554,6 +557,16 @@ public void apply(VariantContext vc, ReadsContext readsContext, ReferenceContext
                 return;
         }
 
+        // Initialize PL index to allele indices for all samples
```
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/SelectVariants.java]: this comment suggests this block should be a method named something like `initializePLindicesToAlleleIndices`

File: src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java [hunk 2]
```diff
@@ -118,7 +109,7 @@ public static GenotypesContext subsetDiploidAlleles(final VariantContext vc,
         if (vc.getGenotypes().isEmpty()) return GenotypesContext.create();
 
         // we need to determine which of the alternate alleles (and hence the likelihoods) to use and carry forward
-        final List<Integer> likelihoodIndexesToUse = determineLikelihoodIndexesToUse(vc, allelesToUse);
+        final List<List<Integer>>likelihoodIndexesToUse = determineLikelihoodIndexesToUse(vc, allelesToUse);
```
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java]: space after `>>`

File: src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java [hunk 3]
```diff
@@ -1017,28 +1004,46 @@ static private GenotypesContext fixGenotypesFromSubsettedAlleles(final Genotypes
...
+        final List<List<Integer>> likelihoodIndexesPerGenotype = new ArrayList<List<Integer>>(10);
 
-        // numLikelihoods takes total # of alleles. Use default # of chromosomes (ploidy) = 2
-        final int numLikelihoods = GenotypeLikelihoods.numLikelihoods(originalVC.getNAlleles(), DEFAULT_PLOIDY);
+        for (final Genotype g : originalVC.getGenotypes()) {
+            final int numLikelihoods = GenotypeLikelihoods.numLikelihoods(originalVC.getNAlleles(), g.getPloidy());
+            final List<Integer> likelihoodIndexes = new ArrayList<Integer>(30);
+            for ( int PLindex = 0; PLindex < numLikelihoods; PLindex++ ) {
+                // consider this entry only if all the alleles are good
+                if ( areAllAllelesUsed(PLindex, alleleIndexesToUse, g.getPloidy()) )
```
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java]: braces

File: src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java [hunk 4]
```diff
@@ -1104,39 +1109,51 @@ private static Genotype fixAD(final Genotype genotype, final boolean[] alleleInd
...
-                                                     final GenotypeBuilder gb,
-                                                     final GenotypeAssignmentMethod assignmentMethod,
-                                                     final double[] newLikelihoods,
-                                                     final List<Allele> allelesToUse) {
+                                                        final int ploidy,
+                                                        final GenotypeBuilder gb,
+                                                        final GenotypeAssignmentMethod assignmentMethod,
+                                                        final double[] newLikelihoods,
+                                                        final List<Allele> allelesToUse) {
+        if ( originalGT == null )
```
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java]: Utils.nonNull everywhere

**Comment by akiezun** [comment]
@ronlevine yes, i just started. will continue later this week

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/SelectVariants.java [hunk 5]
```diff
@@ -1062,7 +1074,9 @@ private VariantContext subsetRecord(final VariantContext vc, final boolean prese
             for (final Genotype genotype : newGC) {
                 //Set genotype to no call if it falls in the fraction.
                 if (fractionGenotypes > 0 && randomGenotypes.nextDouble() < fractionGenotypes) {
-                    genotypes.add(new GenotypeBuilder(genotype).alleles(diploidNoCallAlleles).noGQ().make());
+                    final List<Allele> noCallAlleles = (genotype.getPloidy() == 2 ? diploidNoCallAlleles :
```
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/SelectVariants.java]: the same is duplicated in two places in this file - suggests an extract method is in order: `final List<Allele> noCallAlleles = (g.getPloidy() == 2 ? diploidNoCallAlleles : GATKVariantContextUtils.noCallAlleles(g.getPloidy()));`

File: src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java (hunk 2 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java]: a list of lists is a bad code smell - either document here what this list is or refactor to a proper class

File: src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java [hunk 6]
```diff
@@ -130,21 +121,18 @@ public static GenotypesContext subsetDiploidAlleles(final VariantContext vc,
      * @param originalGs               the original GenotypesContext
      * @param vc                       the original VariantContext
      * @param allelesToUse             the actual alleles to use with the new Genotypes
-     * @param likelihoodIndexesToUse   the indexes in the PL to use given the allelesToUse (@see #determineLikelihoodIndexesToUse())
+     * @param likelihoodIndexesToUse   the indexes in the PL to use given the allelesToUse for each genotype (@see #determineLikelihoodIndexesToUse())
```
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java]: same - document better well or refactor to a special class.

File: src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java [hunk 7]
```diff
@@ -1017,28 +1004,46 @@ static private GenotypesContext fixGenotypesFromSubsettedAlleles(final Genotypes
...
      * @param originalVC           the original VariantContext
      * @param alleleIndexesToUse   the bitset representing the alleles to use (@see #getAlleleIndexBitset)
-     * @return a non-null List
+     * @return likelihoods indexes for each genotype
      */
-    private static List<Integer> getLikelihoodIndexes(final VariantContext originalVC, final boolean[] alleleIndexesToUse) {
+    private static List<List<Integer>> getLikelihoodIndexes(final VariantContext originalVC, final boolean[] alleleIndexesToUse) {
 
-        final List<Integer> result = new ArrayList<>(30);
+        final List<List<Integer>> likelihoodIndexesPerGenotype = new ArrayList<List<Integer>>(10);
```
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java]: all generic allocations should use the diamond type, ie `new ArrayList<>(10);`

File: src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java [hunk 8]
```diff
@@ -1017,28 +1004,46 @@ static private GenotypesContext fixGenotypesFromSubsettedAlleles(final Genotypes
...
+
+    /**
+     * For a PL index, check if all of the alleles are used
+     *
+     * @param PLindex               index into the number of likelihoods (@see #numLikelihoods)
+     * @param alleleIndexesToUse    the bitset representing the alleles to use (@see #getAlleleIndexBitset)
+     * @param ploidy                the number of chromosomes
+     * @return true if all of the alleles are used, false otherwise
+     */
+    private static boolean areAllAllelesUsed(final int PLindex, final boolean[] alleleIndexesToUse, final int ploidy) {
```
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java]: replace this whole method by a simple `GenotypeLikelihoods.getAlleles(PLindex, ploidy).stream().allMatch(i -> alleleIndexesToUse[i]);`

**Comment by akiezun** [comment]
back to @ronlevine - can you address the comments and rebase? it's like 40 commits behind master.

File: src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java (hunk 2 above)
Comment (ronlevine) [inline src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java]: Changed the name to `genpotypesLikelihoodIndexesToUse` and improved the documentation. For this case, the list of lists seems natural to me.

File: src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java (hunk 6 above)
Comment (ronlevine) [inline src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java]: Done.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/SelectVariants.java (hunk 5 above)
Comment (ronlevine) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/SelectVariants.java]: Done

File: src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java (hunk 7 above)
Comment (ronlevine) [inline src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java]: Done.

File: src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java (hunk 8 above)
Comment (ronlevine) [inline src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java]: Done. Was stuck in in Java 7 and...

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/SelectVariants.java (hunk 1 above)
Comment (ronlevine) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/SelectVariants.java]: Changed to `Initialize the cache of PL index to a list of alleles for each ploidy.`

File: src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java (hunk 3 above)
Comment (ronlevine) [inline src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java]: Done.

File: src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java (hunk 4 above)
Comment (ronlevine) [inline src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java]: Done.

**Comment by ronlevine** [comment]
back to @akiezun. addressed comments and rebased.

**Comment by akiezun** [comment]
@ronlevine is this ready? there are new commits all the time.

**Comment by ronlevine** [comment]
It's ready.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/SelectVariants.java [hunk 9]
```diff
@@ -554,6 +557,18 @@ public void apply(VariantContext vc, ReadsContext readsContext, ReferenceContext
                 return;
         }
 
+        // Initialize the cache of PL index to a list of alleles for each ploidy.
```
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/SelectVariants.java]: extract this to a new method (i think my previous comments were wiped out btw)

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/SelectVariants.java [hunk 10]
```diff
@@ -437,7 +438,9 @@
     private Set<String> IDsToKeep = null;
     private Set<String> IDsToRemove = null;
 
-    private final List<Allele> diploidNoCallAlleles = Arrays.asList(Allele.NO_CALL, Allele.NO_CALL);
+    private final List<Allele> diploidNoCallAlleles = GATKVariantContextUtils.noCallAlleles(2);
+
+    private final Map<Integer, Integer> ploidyToNumberOfAlleles = new HashMap<Integer, Integer>();
```
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/SelectVariants.java]: dont use HashMap - either treemap or linkedhashmap. We need stable iteration order

File: src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java [hunk 11]
```diff
@@ -68,37 +68,29 @@ public static VariantContextWriter createVCFWriter(
...
         if ( allelesToUse.get(0).isNonReference() ) throw new IllegalArgumentException("First allele must be the reference allele");
         if ( allelesToUse.size() == 1 ) throw new IllegalArgumentException("Cannot subset to only 1 alt allele");
 
         // optimization: if no input genotypes, just exit
         if (vc.getGenotypes().isEmpty()) return GenotypesContext.create();
 
-        // we need to determine which of the alternate alleles (and hence the likelihoods) to use and carry forward
-        final List<Integer> likelihoodIndexesToUse = determineLikelihoodIndexesToUse(vc, allelesToUse);
+        // from a subset of alternate alleles to use, determine the likelihood indexes to use for each genotype
+        final List<List<Integer>> genpotypesLikelihoodIndexesToUse = determineGenotypesLikelihoodIndexesToUse(vc, allelesToUse);
```
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java]: typo `genpotypes...`

File: src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java [hunk 12]
```diff
@@ -107,21 +99,18 @@ public static GenotypesContext subsetDiploidAlleles(final VariantContext vc,
      * @param originalGs               the original GenotypesContext
      * @param vc                       the original VariantContext
      * @param allelesToUse             the actual alleles to use with the new Genotypes
-     * @param likelihoodIndexesToUse   the indexes in the PL to use given the allelesToUse (@see #determineLikelihoodIndexesToUse())
+     * @param genpotypesLikelihoodIndexesToUse   for each genotype, the indexes in the PL to use given the allelesToUse (@see #determineGenotypesLikelihoodIndexesToUse())
```
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java]: usual plural of index is indices

File: src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java [hunk 13]
```diff
@@ -964,22 +952,22 @@ public static GenotypesContext updatePLsAndAD(final VariantContext selectedVC, f
      * @param allelesToUse     the new (sub)set of alleles to use
      * @return a new non-null GenotypesContext
      */
-    static private GenotypesContext fixGenotypesFromSubsettedAlleles(final GenotypesContext originalGs, final VariantContext originalVC, final List<Allele> allelesToUse) {
+    private static GenotypesContext fixGenotypesFromSubsettedAlleles(final GenotypesContext originalGs, final VariantContext originalVC, final List<Allele> allelesToUse) {
 
         // we need to determine which of the alternate alleles (and hence the likelihoods) to use and carry forward
-        final List<Integer> likelihoodIndexesToUse = determineLikelihoodIndexesToUse(originalVC, allelesToUse);
+        final List<List<Integer>> genpotypesLikelihoodIndexesToUse = determineGenotypesLikelihoodIndexesToUse(originalVC, allelesToUse);
```
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java]: typo `genpotypes`

File: src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java [hunk 14]
```diff
@@ -964,22 +952,22 @@ public static GenotypesContext updatePLsAndAD(final VariantContext selectedVC, f
...
+        return createGenotypesWithSubsettedLikelihoods(originalGs, originalVC, allelesToUse, genpotypesLikelihoodIndexesToUse, GenotypeAssignmentMethod.DO_NOT_ASSIGN_GENOTYPES);
     }
     /**
-     * Figure out which likelihood indexes to use for a selected down set of alleles
+     * Determine which likelihood indexes to use for each genotype for a subset of alleles
      *
      * @param originalVC        the original VariantContext
      * @param allelesToUse      the subset of alleles to use
-     * @return a list of PL indexes to use or null if none
+     * @return likelihood indexes for each genotype
```
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java]: this comment does not correspond to the type of the parameter

Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java]: this method needs better documentation if what the returned value is - what is this list of lists of?

File: src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java [hunk 15]
```diff
@@ -994,28 +982,28 @@ static private GenotypesContext fixGenotypesFromSubsettedAlleles(final Genotypes
...
-
-        for ( int PLindex = 0; PLindex < numLikelihoods; PLindex++ ) {
-            final GenotypeLikelihoods.GenotypeLikelihoodsAllelePair alleles = GenotypeLikelihoods.getAllelePair(PLindex);
-            // consider this entry only if both of the alleles are good
-            if ( alleleIndexesToUse[alleles.alleleIndex1] && alleleIndexesToUse[alleles.alleleIndex2] ) {
-                result.add(PLindex);
+        for (final Genotype g : originalVC.getGenotypes()) {
+            final int numLikelihoods = GenotypeLikelihoods.numLikelihoods(originalVC.getNAlleles(), g.getPloidy());
+            final List<Integer> likelihoodIndexes = new ArrayList<Integer>(30);
+            for ( int PLindex = 0; PLindex < numLikelihoods; PLindex++ ) {
```
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java]: rename this variable. temp variables are named in lowercase

File: src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java [hunk 16]
```diff
@@ -994,28 +982,28 @@ static private GenotypesContext fixGenotypesFromSubsettedAlleles(final Genotypes
...
-        final int numLikelihoods = GenotypeLikelihoods.numLikelihoods(originalVC.getNAlleles(), DEFAULT_PLOIDY);
-
-        for ( int PLindex = 0; PLindex < numLikelihoods; PLindex++ ) {
-            final GenotypeLikelihoods.GenotypeLikelihoodsAllelePair alleles = GenotypeLikelihoods.getAllelePair(PLindex);
-            // consider this entry only if both of the alleles are good
-            if ( alleleIndexesToUse[alleles.alleleIndex1] && alleleIndexesToUse[alleles.alleleIndex2] ) {
-                result.add(PLindex);
+        for (final Genotype g : originalVC.getGenotypes()) {
+            final int numLikelihoods = GenotypeLikelihoods.numLikelihoods(originalVC.getNAlleles(), g.getPloidy());
+            final List<Integer> likelihoodIndexes = new ArrayList<Integer>(30);
```
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java]: use diamonds: `new ArrayList<>(30)`

File: src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java [hunk 17]
```diff
@@ -1081,39 +1069,49 @@ private static Genotype fixAD(final Genotype genotype, final boolean[] alleleInd
...
-                    gb.alleles(noCallAlleles(2)); // @TODO assumes diploid
+                    gb.alleles(noCallAlleles(ploidy));
                     gb.noGQ();
                 } else {
                     // find the genotype with maximum likelihoods
                     final int PLindex = MathUtils.maxElementIndex(newLikelihoods);
-                    GenotypeLikelihoods.GenotypeLikelihoodsAllelePair alleles = GenotypeLikelihoods.getAllelePair(PLindex);
-                    gb.alleles(Arrays.asList(allelesToUse.get(alleles.alleleIndex1), allelesToUse.get(alleles.alleleIndex2)));
+                    final List<Allele> alleles = new ArrayList<>();
+                    for ( final Integer alleleIndex : GenotypeLikelihoods.getAlleles(PLindex, ploidy)) {
```
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java]: use `int` here

**Comment by akiezun** [comment]
back to @ronlevine to address the comments. Small things. Once you address, rebase and squash and we'll merge

**Comment by ronlevine** [comment]
@akiezun Comments addressed, rebaseed and squashed.

**Comment by akiezun** [comment]
LGTM

**Pull request merged**